File: mobile_restaurant/utils/Validators.ts
/**
 * Centralized validators for form fields.
 * Synced with backend RegisterRequest.java validation rules:
 * - username: @NotBlank, @Length(min=3)
 * - email: @NotBlank, @Email
 * - phone: @NotBlank, @Pattern("^(0[35789][0-9]{8})$")
 * - password: @NotBlank, @Length(min=8), @Pattern(".*[A-Z].*")
 * - confirmPassword: @NotBlank, must match password
 */
const EMAIL_REGEX = /^[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}$/;
const PHONE_REGEX = /^(0[35789][0-9]{8})$/;

export const Validators = {
  email(value?: string, fieldName = 'Email'): string | null {
    if (!value || !value.trim()) return `${fieldName} không được để trống`;
    if (!EMAIL_REGEX.test(value.trim())) return `Vui lòng nhập đúng định dạng email`;
    return null;
  },

  password(
    value?: string,
    options: { minLength?: number; fieldName?: string } = {},
  ): string | null {
    const { minLength = 8, fieldName = 'Mật khẩu' } = options;
    if (!value || !value.trim()) return `${fieldName} không được để trống`;
    const v = value.trim();
    if (v.length < minLength) return `Độ dài mật khẩu ít nhất ${minLength} ký tự`;
    if (!/[A-Z]/.test(v)) return `Mật khẩu phải chứa ít nhất một ký tự in hoa`;
    return null;
  },

  confirmPassword(getPassword: () => string, fieldName = 'Mật khẩu xác nhận') {
    return (value?: string): string | null => {
      if (!value || !value.trim()) return `${fieldName} không được để trống`;
      if (value.trim() !== getPassword().trim()) return 'Mật khẩu nhập lại không khớp';
      return null;
    };
  },

  phone(value?: string, fieldName = 'Số điện thoại'): string | null {
    if (!value || !value.trim()) return `${fieldName} không được để trống`;
    if (!PHONE_REGEX.test(value.trim())) return `Vui lòng nhập đúng định dạng số điện thoại`;
    return null;
  },

  required(value?: string, fieldName = 'Trường này'): string | null {
    if (!value || !value.trim()) return `${fieldName} không được để trống`;
    return null;
  },

  name(value?: string, fieldName = 'Tên'): string | null {
    if (!value || !value.trim()) return `${fieldName} không được để trống`;
    if (value.trim().length < 3) return `Độ dài của tên ít nhất 3 ký tự`;
    return null;
  },

  otp(value?: string, length = 6, fieldName = 'Mã OTP'): string | null {
    if (!value || !value.trim()) return `${fieldName} không được để trống`;
    if (value.trim().length !== length) return `${fieldName} phải ${length} số`;
    return null;
  },
};

File: lib/Utils/Validators.dart
/// Centralized validators for form fields.
class Validators {
  Validators._(); // Prevent instantiation

  static final RegExp _emailRegex = RegExp(r'^[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}$');

  // ==================== Email ====================
  static String? email(String? value, {String? fieldName}) {
    final name = fieldName ?? 'Email';
    if (value == null || value.trim().isEmpty) {
      return 'Vui long nhap $name';
    }
    if (!_emailRegex.hasMatch(value.trim())) {
      return '$name khong hop le';
    }
    return null;
  }

  // ==================== Password ====================
  static String? password(String? value, {int minLength = 6, String? fieldName, bool requireUppercase = false}) {
    final name = fieldName ?? 'Mat khau';
    if (value == null || value.trim().isEmpty) {
      return 'Vui long nhap $name';
    }
    final v = value.trim();
    if (v.length < minLength) {
      return '$name toi thieu $minLength ky tu';
    }
    if (requireUppercase && !RegExp(r'[A-Z]').hasMatch(v)) {
      return '$name phai co it nhat 1 ky tu in hoa';
    }
    return null;
  }

  /// Validator cho xac nhan mat khau
  static String? Function(String?) confirmPassword(
    String Function() getPassword, {
    String? fieldName,
  }) {
    return (String? value) {
      final name = fieldName ?? 'Xac nhan mat khau';
      if (value == null || value.trim().isEmpty) {
        return 'Vui long nhap $name';
      }
      if (value.trim() != getPassword().trim()) {
        return 'Mat khau nhap lai khong khop';
      }
      return null;
    };
  }

  // ==================== Phone ====================
  static String? phone(String? value, {String? fieldName}) {
    final name = fieldName ?? 'So dien thoai';
    if (value == null || value.trim().isEmpty) {
      return 'Vui long nhap $name';
    }
    final cleaned = value.trim().replaceAll(RegExp(r'[^0-9]'), '');
    if (cleaned.length < 9 || cleaned.length > 11) {
      return '$name khong hop le';
    }
    return null;
  }

  // ==================== Required ====================
  static String? required(String? value, {String? fieldName}) {
    final name = fieldName ?? 'Truong nay';
    if (value == null || value.trim().isEmpty) {
      return 'Vui long nhap $name';
    }
    return null;
  }

  // ==================== Name ====================
  static String? name(String? value, {String? fieldName}) {
    final name = fieldName ?? 'Ho ten';
    if (value == null || value.trim().isEmpty) {
      return 'Vui long nhap $name';
    }
    if (value.trim().length < 2) {
      return '$name qua ngan';
    }
    return null;
  }

  // ==================== OTP ====================
  static String? otp(String? value, {int length = 6, String? fieldName}) {
    final name = fieldName ?? 'Ma OTP';
    if (value == null || value.trim().isEmpty) {
      return 'Vui long nhap $name';
    }
    if (value.trim().length != length) {
      return '$name phai $length so';
    }
    return null;
  }

  // ==================== Min Length ====================
  static String? Function(String?) minLength(int min, {String? fieldName}) {
    return (String? value) {
      final name = fieldName ?? 'Truong nay';
      if (value == null || value.trim().isEmpty) {
        return 'Vui long nhap $name';
      }
      if (value.trim().length < min) {
        return '$name toi thieu $min ky tu';
      }
      return null;
    };
  }

  // ==================== Combine Validators ====================
  /// Ket hop nhieu validators thanh 1
  static String? Function(String?) combine(List<String? Function(String?)> validators) {
    return (String? value) {
      for (final validator in validators) {
        final error = validator(value);
        if (error != null) return error;
      }
      return null;
    };
  }
}
